File: packages/observability/typescript/src/logger.ts
/**
 * MIT License
 *
 * Copyright (c) 2026 RocketRide, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Structured JSON Logger with OpenTelemetry trace context support.
 *
 * Writes JSON lines to stdout (info/debug) or stderr (warn/error).
 * Supports child loggers that inherit and extend context.
 */

import { scrubPii } from './pii';

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

const LEVEL_PRIORITY: Record<LogLevel, number> = {
	debug: 10,
	info: 20,
	warn: 30,
	error: 40,
};

export interface LoggerOptions {
	/** Minimum log level to emit. Defaults to 'info'. */
	level?: LogLevel;
	/** Enable PII scrubbing. Defaults to true. */
	scrubPii?: boolean;
	/** Custom output function for testing. Defaults to process.stdout/stderr.write. */
	output?: (json: string, level: LogLevel) => void;
}

function getOtelContext(): { trace_id: string; span_id: string; trace_flags: string } {
	try {
		// eslint-disable-next-line @typescript-eslint/no-var-requires
		const api = require('@opentelemetry/api');
		const span = api.trace.getActiveSpan?.();
		if (span) {
			const ctx = span.spanContext();
			if (ctx && ctx.traceId && ctx.traceId !== '00000000000000000000000000000000') {
				return {
					trace_id: ctx.traceId,
					span_id: ctx.spanId,
					trace_flags: ctx.traceFlags?.toString(16).padStart(2, '0') ?? '',
				};
			}
		}
	} catch {
		// OTel not available — that's fine
	}
	return { trace_id: '', span_id: '', trace_flags: '' };
}

function defaultOutput(json: string, level: LogLevel): void {
	if (level === 'error' || level === 'warn') {
		process.stderr.write(json + '\n');
	} else {
		process.stdout.write(json + '\n');
	}
}

export class Logger {
	private _name: string;
	private _context: Record<string, unknown>;
	private _level: LogLevel;
	private _scrub: boolean;
	private _output: (json: string, level: LogLevel) => void;

	constructor(name: string, context: Record<string, unknown> = {}, options: LoggerOptions = {}) {
		this._name = name;
		this._context = { ...context };
		this._level = options.level ?? 'info';
		this._scrub = options.scrubPii ?? true;
		this._output = options.output ?? defaultOutput;
	}

	/**
	 * Create a child logger that inherits this logger's context
	 * and adds additional bound fields.
	 */
	child(context: Record<string, unknown>): Logger {
		const merged = { ...this._context, ...context };
		return new Logger(this._name, merged, {
			level: this._level,
			scrubPii: this._scrub,
			output: this._output,
		});
	}

	debug(event: string, data?: Record<string, unknown>): void {
		this._log('debug', event, data);
	}

	info(event: string, data?: Record<string, unknown>): void {
		this._log('info', event, data);
	}

	warn(event: string, data?: Record<string, unknown>): void {
		this._log('warn', event, data);
	}

	error(event: string, data?: Record<string, unknown>): void {
		this._log('error', event, data);
	}

	private _log(level: LogLevel, event: string, data?: Record<string, unknown>): void {
		if (LEVEL_PRIORITY[level] < LEVEL_PRIORITY[this._level]) {
			return;
		}

		const otel = getOtelContext();

		let record: Record<string, unknown> = {
			timestamp: new Date().toISOString(),
			level,
			logger: this._name,
			event,
			...otel,
			...this._context,
			...data,
		};

		if (this._scrub) {
			record = scrubPii(record);
		}

		this._output(JSON.stringify(record), level);
	}
}

/**
 * Create a new structured logger.
 *
 * @param name - Logger name (typically the module or service name)
 * @param options - Logger configuration
 */
export function createLogger(name: string, options?: LoggerOptions): Logger {
	return new Logger(name, {}, options);
}

File: packages/observability/python/rocketride_observability/logger.py
# MIT License
#
# Copyright (c) 2026 RocketRide, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""Structured logging built on structlog wrapping stdlib logging.

Provides JSON-formatted log output with OpenTelemetry trace context
injection and PII scrubbing.
"""

import logging
import sys

import structlog

from .pii import scrub_pii_processor

_configured = False


def _add_otel_context(logger: logging.Logger, method_name: str, event_dict: dict) -> dict:
    """Inject OpenTelemetry trace and span IDs into the log event."""
    try:
        from opentelemetry import trace

        span = trace.get_current_span()
        ctx = span.get_span_context()
        if ctx and ctx.trace_id:
            event_dict['trace_id'] = format(ctx.trace_id, '032x')
            event_dict['span_id'] = format(ctx.span_id, '016x')
            event_dict['trace_flags'] = format(ctx.trace_flags, '02x')
        else:
            event_dict.setdefault('trace_id', '')
            event_dict.setdefault('span_id', '')
    except Exception:
        event_dict.setdefault('trace_id', '')
        event_dict.setdefault('span_id', '')
    return event_dict


def configure_logging(level: int = logging.INFO) -> None:
    """Configure structlog and stdlib logging for JSON output.

    Call once at application startup. Subsequent calls are no-ops.
    """
    global _configured
    if _configured:
        return

    shared_processors = [
        structlog.contextvars.merge_contextvars,
        structlog.stdlib.add_log_level,
        structlog.processors.TimeStamper(fmt='iso'),
        _add_otel_context,
        scrub_pii_processor,
    ]

    structlog.configure(
        processors=[
            *shared_processors,
            structlog.stdlib.ProcessorFormatter.wrap_for_formatter,
        ],
        wrapper_class=structlog.stdlib.BoundLogger,
        context_class=dict,
        logger_factory=structlog.stdlib.LoggerFactory(),
        cache_logger_on_first_use=True,
    )

    formatter = structlog.stdlib.ProcessorFormatter(
        processors=[
            structlog.stdlib.ProcessorFormatter.remove_processors_meta,
            structlog.processors.JSONRenderer(),
        ],
    )

    handler = logging.StreamHandler(sys.stderr)
    handler.setFormatter(formatter)

    root = logging.getLogger()
    root.handlers.clear()
    root.addHandler(handler)
    root.setLevel(level)

    _configured = True


def get_logger(name: str, **initial_context: object) -> structlog.stdlib.BoundLogger:
    """Return a structured logger bound with the given name and optional context.

    Automatically calls configure_logging() if it hasn't been called yet,
    so callers can simply use get_logger() without explicit setup.
    """
    if not _configured:
        configure_logging()
    return structlog.get_logger(name, **initial_context)
